### DISCUSSION TITLE
vercel dev errror: 'cmd.exe' is not recognized as an internal or external command

### DISCUSSION THREAD

**Discussion by faderik** [vercel/vercel#8762] [Help]
I am new & firstly using `vercel dev`. but when running vercel dev on windows through GitBash, getting this error

```bash
$ vercel dev
Vercel CLI 28.4.12
> Creating initial build
'cmd.exe' is not recognized as an internal or external command,
    at ChildProcess.cp.emit (C:\Users\user\AppData\Local\Yarn\Data\global\node_modules\@vercel\build-utils\dist\index.js:17909:37)
    at Process.ChildProcess._handle.onexit (node:internal/child_process:290:12) {
  code: 'ENOENT',
  errno: 'ENOENT',
  syscall: 'spawn cmd.exe',
  path: 'cmd.exe',
  spawnargs: [ '/C', 'yarn run dev' ]
}
Error: Builder exited with 1 before sending build result
```

Did I do somthing wrong?

also found [this](https://stackoverflow.com/questions/32035577/cmd-is-not-recognized-as-an-internal-or-external-command-operable-program-or) similiar issue, but nothing works.

**Comment by amyegan** [comment]
I don't think you did anything wrong @faderik. There are some differences between Windows and Linux/Mac environments that can make CLI tools a bit difficult to configure at times. Do you get the same error from other terminal options, such as Command Prompt, PowerShell, or [Windows Terminal](https://github.com/Microsoft/Terminal)?

**Reply by faderik** [reply]
Yes, I got same error with all terminal I have on Windows

![image](https://user-images.githubusercontent.com/55375390/197928099-f8950793-e92d-4996-a2da-dde01370df75.png)

**Comment by Nexum** [comment]
Having the same problem in webstorm when using a run config @28.7.0

**Reply by amyegan** [reply]
Have you already tried the Stack Overflow suggestions that faderik shared? 
- [Check that `system32` is correctly defined in `PATH`](https://stackoverflow.com/a/32035723)
- Make sure you [have `PATH` and not just `Path` set](https://stackoverflow.com/a/54664542), as capitalization may make a difference

**Comment by MacaylaMarvelous81** [comment]
I'm running into the same issue on Windows Terminal using Git Bash. Through the terminal I can run `cmd` and `cmd.exe` but strangely vercel CLI fails:

```bash
Vercel CLI 28.8.0
> Running Dev Command “next dev --port $PORT”
'cmd.exe' is not recognized as an internal or external command,
operable program or batch file.
Error: spawn cmd.exe ENOENT
    at notFoundError (C:\Users\ilove\Appdata\Local\Yarn\Data\global\node_modules\vercel\dist\index.js:177999:26)
    at verifyENOENT (C:\Users\ilove\Appdata\Local\Yarn\Data\global\node_modules\vercel\dist\index.js:178033:16)
    at ChildProcess.cp.emit (C:\Users\ilove\Appdata\Local\Yarn\Data\global\node_modules\vercel\dist\index.js:178020:25)
    at Process.ChildProcess._handle.onexit (node:internal/child_process:291:12)
Error: An unexpected error occurred!
Error: spawn cmd.exe ENOENT
    at notFoundError (C:\Users\ilove\Appdata\Local\Yarn\Data\global\node_modules\vercel\dist\index.js:177999:26)
    at verifyENOENT (C:\Users\ilove\Appdata\Local\Yarn\Data\global\node_modules\vercel\dist\index.js:178033:16)
    at ChildProcess.cp.emit (C:\Users\ilove\Appdata\Local\Yarn\Data\global\node_modules\vercel\dist\index.js:178020:25)
    at Process.ChildProcess._handle.onexit (node:internal/child_process:291:12)
```

Through yarn the command it tries to run works fine.
```bash
yarn run next dev --port 3000
yarn run v1.22.19
$ C:\Users\ilove\Documents\macaylamarvelous81-blog\node_modules\.bin\next dev --port 3000
ready - started server on 0.0.0.0:3000, url: http://localhost:3000
```

**Reply by amyegan** [reply]
Sorry you're having trouble with this @MacaylaMarvelous81. Can you run `echo %path%` and confirm that the output includes `C:\windows\system32`?

**Reply by MacaylaMarvelous81** [reply]
Yes, it does. vercel cli succeeds in command prompt(cmd):
```bat
Vercel CLI 28.8.0
> Running Dev Command “next dev --port $PORT”
ready - started server on 0.0.0.0:3000, url: http://localhost:3000
> Ready! Available at http://localhost:3000
event - compiled client and server successfully in 652 ms (161 modules)
```
in git bash, vercel cli fails but also includes the system32 in PATH

**Comment by NALZA** [comment]
Id like to add that I am also running into this issue, only while using git bash

**Reply by amyegan** [reply]
I just tried it again on a Windows 10 computer and could not reproduce the error. I found another thread with some good details and tips on debugging `"is not recognized as an internal or external command"` errors. Maybe some of these tricks will help you solve this problem: https://stackoverflow.com/questions/48321639/error-is-not-recognized-as-an-internal-or-external-command-operable-program-o

**Reply by NALZA** [reply]
Hi, @amyegan  I have found where the path is being broken when using the cli. 

Using version 28.10.2 on line :247307 of the index file _vercel/dist/index.js_ where the path is being modified to add node_modules/.bin, **Env.PATH is undefined**. 

I am currently unsure where or how this ends up undefined but I can see the "ORIGINAL_PATH" is set to the correct path. 

Adding the path variables locations for nodejs, npm and system32 on this line I was also able to have vercel dev run in git bash

**Reply by NALZA** [reply]
After a little more in the function cloneEnv, in git bash obj.Path is undefined but obj.PATH is set to the correct path so when obj.PATH is set to obj.Path its set to undefined.

The opposite is true for other windows terminals and the correct obj.PATH is created. 

Adding this check to the function I was able to have it work in both cmd and git bash

 ```
 if(obj.Path !== undefined){
      obj.PATH = obj.Path;

       delete obj.Path;
}
```

I have created a PR. https://github.com/vercel/vercel/pull/9163

**Reply by amyegan** [reply]
Thanks for finding this @NALZA! I asked the team to take a look 👀

**Comment by Ipicon** [comment]
I couldn't figure the "right way" to do it for a whole day,
what I ended up doing is creating a file named 'cmd.exe'.
I ran this command in cmd:
`type "C:\Windows\system32\cmd.exe" > cmd.exe`

**Comment by Ipicon** [comment]
I noticed that this issue only occurred after I installed WSL, maybe it has something to do with that...
`vercel dev` doesn't work on any of my windows terminals without the small fix I provided above.
but  on my WSL machine it works just fine.

**Reply by MacaylaMarvelous81** [reply]
Possibly! I've had WSL installed to utilize Docker and just a few days ago I switched to using WSL instead of Git Bash (though I haven't tried to use `vercel dev` since)
